hazelcast-wm/src/main/java/com/hazelcast/web/WebFilter.java

```diff
@@ -475,14 +476,21 @@ private HazelcastHttpSession readSessionFromLocal() {
                 String hazelcastSessionId = originalSessions.get(originalSession.getId());
                 if (hazelcastSessionId != null) {
                     hazelcastSession = sessions.get(hazelcastSessionId);
-
                     if (hazelcastSession != null && !hazelcastSession.isStickySession()) {
                         hazelcastSession.updateReloadFlag();
                     }
                     return hazelcastSession;
                 }
-                originalSessions.remove(originalSession.getId());
-                originalSession.invalidate();
+                // Even though session can be taken from request, it might be already invalidated.
+                // For example, in Wildfly (uses Undertow), taken wrapper session might be valid
+                // but its underlying real session might be already invalidated after redirection
+                // due to its request/url based wrapper session (points to same original session) design.
+                // Therefore, we check the taken session id and
+                // ignore its invalidation if it is already invalidated inside Hazelcast's session.
+                // See issue on Wildfly https://github.com/hazelcast/hazelcast/issues/6335
+                if (!originalSession.getId().equals(invalidatedOriginalSessionId)) {
+                    originalSession.invalidate();
```

Comment:
when code reached here it is obvious that `originalSession.getId()` is not in `originalSessions` so it was redundant and removed it.
Check this code snippet before here:
``` java
String hazelcastSessionId = originalSessions.get(originalSession.getId());
if (hazelcastSessionId != null) {
    ...
    return
}
```